File: src/apis/knowledge.ts
import type { IndexResult, KnowledgeBase, KnowledgeDetail, RetrievalSource } from '@/types';

export async function fetchKnowledgeList(): Promise<KnowledgeBase[]> {
  const res = await fetch('/api/knowledge');
  if (!res.ok) {
    throw new Error(`Knowledge request failed: HTTP ${res.status}`);
  }
  const data = (await res.json()) as { knowledge: KnowledgeBase[] };
  return data.knowledge;
}

export async function createKnowledge(name: string, description = ''): Promise<KnowledgeBase> {
  const res = await fetch('/api/knowledge', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name, description }),
  });
  if (!res.ok) {
    throw new Error(`Create knowledge failed: HTTP ${res.status}`);
  }
  return (await res.json()) as KnowledgeBase;
}

export async function fetchKnowledgeDetail(id: string): Promise<KnowledgeDetail> {
  const res = await fetch(`/api/knowledge/${encodeURIComponent(id)}`);
  if (!res.ok) {
    throw new Error(`Knowledge detail failed: HTTP ${res.status}`);
  }
  return (await res.json()) as KnowledgeDetail;
}

export async function updateKnowledge(
  id: string,
  payload: { name?: string; description?: string }
): Promise<KnowledgeBase> {
  const res = await fetch(`/api/knowledge/${encodeURIComponent(id)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error(`Update knowledge failed: HTTP ${res.status}`);
  }
  return (await res.json()) as KnowledgeBase;
}

export async function deleteKnowledge(id: string): Promise<void> {
  const res = await fetch(`/api/knowledge/${encodeURIComponent(id)}`, { method: 'DELETE' });
  if (!res.ok) {
    throw new Error(`Delete knowledge failed: HTTP ${res.status}`);
  }
}

export async function bindKnowledgeFile(knowledgeId: string, fileId: string): Promise<KnowledgeDetail> {
  const res = await fetch(`/api/knowledge/${encodeURIComponent(knowledgeId)}/files`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ file_id: fileId }),
  });
  if (!res.ok) {
    throw new Error(`Bind file failed: HTTP ${res.status}`);
  }
  return (await res.json()) as KnowledgeDetail;
}

export async function unbindKnowledgeFile(knowledgeId: string, fileId: string): Promise<KnowledgeDetail> {
  const res = await fetch(
    `/api/knowledge/${encodeURIComponent(knowledgeId)}/files/${encodeURIComponent(fileId)}`,
    { method: 'DELETE' }
  );
  if (!res.ok) {
    throw new Error(`Unbind file failed: HTTP ${res.status}`);
  }
  return (await res.json()) as KnowledgeDetail;
}

export async function indexKnowledge(id: string): Promise<IndexResult> {
  const res = await fetch(`/api/knowledge/${encodeURIComponent(id)}/index`, { method: 'POST' });
  if (!res.ok) {
    let detail = `HTTP ${res.status}`;
    try {
      const body = await res.json();
      detail = body.detail ?? detail;
    } catch { /* ignore */ }
    throw new Error(`建立索引失败：${detail}`);
  }
  return (await res.json()) as IndexResult;
}

export async function queryKnowledge(
  id: string,
  query: string,
  topK = 4
): Promise<RetrievalSource[]> {
  const res = await fetch(`/api/knowledge/${encodeURIComponent(id)}/query`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ query, top_k: topK }),
  });
  if (!res.ok) {
    throw new Error(`Query knowledge failed: HTTP ${res.status}`);
  }
  const data = (await res.json()) as { chunks: Array<Record<string, unknown>> };
  return (data.chunks ?? []).map((chunk, idx) => ({
    index: idx + 1,
    file_id: String(chunk.file_id ?? ''),
    filename: String(chunk.filename ?? ''),
    chunk_index: Number(chunk.chunk_index ?? 0),
    score: Number(chunk.score ?? 0),
    preview: String(chunk.content ?? ''),
  }));
}

File: server/routers/knowledge.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Response, status

from server.repositories.configs import get_provider_config
from server.repositories.knowledge import (
    bind_file,
    create_knowledge,
    delete_knowledge,
    get_knowledge,
    knowledge_exists,
    list_knowledge,
    unbind_file,
    update_knowledge,
)
from server.schemas.knowledge import (
    BindFileRequest,
    IndexResult,
    Knowledge,
    KnowledgeCreate,
    KnowledgeDetail,
    KnowledgeListResponse,
    KnowledgeUpdate,
    RetrievalChunk,
    RetrievalQuery,
    RetrievalResult,
)
from server.services.rag import index_knowledge, query_knowledge

router = APIRouter(prefix="/api/knowledge", tags=["knowledge"])


@router.get("", response_model=KnowledgeListResponse)
def get_knowledge_list() -> KnowledgeListResponse:
    return KnowledgeListResponse(knowledge=list_knowledge())


@router.post("", response_model=Knowledge)
def post_knowledge(payload: KnowledgeCreate) -> Knowledge:
    return create_knowledge(payload.name, payload.description)


@router.get("/{knowledge_id}", response_model=KnowledgeDetail)
def get_knowledge_detail(knowledge_id: str) -> KnowledgeDetail:
    detail = get_knowledge(knowledge_id)
    if not detail:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    return detail


@router.put("/{knowledge_id}", response_model=Knowledge)
def put_knowledge(knowledge_id: str, payload: KnowledgeUpdate) -> Knowledge:
    updated = update_knowledge(knowledge_id, payload.name, payload.description)
    if not updated:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    return updated


@router.delete("/{knowledge_id}", status_code=status.HTTP_204_NO_CONTENT)
def remove_knowledge(knowledge_id: str) -> Response:
    delete_knowledge(knowledge_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{knowledge_id}/files", response_model=KnowledgeDetail)
def bind_knowledge_file(knowledge_id: str, payload: BindFileRequest) -> KnowledgeDetail:
    ok = bind_file(knowledge_id, payload.file_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Knowledge base or file not found")
    detail = get_knowledge(knowledge_id)
    assert detail is not None
    return detail


@router.delete("/{knowledge_id}/files/{file_id}", response_model=KnowledgeDetail)
def unbind_knowledge_file(knowledge_id: str, file_id: str) -> KnowledgeDetail:
    if not knowledge_exists(knowledge_id):
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    unbind_file(knowledge_id, file_id)
    detail = get_knowledge(knowledge_id)
    assert detail is not None
    return detail


@router.post("/{knowledge_id}/index", response_model=IndexResult)
async def index_knowledge_base(knowledge_id: str) -> IndexResult:
    if not knowledge_exists(knowledge_id):
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    config = get_provider_config()
    result = await index_knowledge(config, config.embedding_model, knowledge_id)
    return IndexResult(embedding_model=config.embedding_model, **result)


@router.post("/{knowledge_id}/query", response_model=RetrievalResult)
async def query_knowledge_base(knowledge_id: str, payload: RetrievalQuery) -> RetrievalResult:
    if not knowledge_exists(knowledge_id):
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    config = get_provider_config()
    chunks = await query_knowledge(
        config,
        config.embedding_model,
        knowledge_id,
        payload.query,
        payload.top_k,
        mode=payload.mode,
        rerank=payload.rerank,
    )
    return RetrievalResult(chunks=[RetrievalChunk(**chunk) for chunk in chunks])
